lib/order_pdf.rb

```diff
@@ -1,104 +1,130 @@
+    down_or_page 25
+    text name, size: 10, style: :bold
+    table data, width: bounds.width, cell_style: {size: 8, overflow: :shrink_to_fit} do |table|
+      # borders
+      table.cells.borders = [:bottom]
+      table.cells.padding_top = 2
+      table.cells.padding_bottom = 4
+      table.cells.border_color = 'dddddd'
+      table.rows(0).border_color = '666666'
+
+      # dim rows which were ordered but not received
+      dimrows.each do |ri|
+        table.row(ri).text_color = '999999'
+        table.row(ri).columns(0..-1).font_style = nil
       end
+
+      yield table if block_given?
     end
+  end
 
-    body  # Add content, which is defined in subclasses
+  protected
 
-    render  # Render pdf
+  # Return price for order_article.
+  #
+  # This is a separate method so that plugins can override it.
+  #
+  # @param article [OrderArticle]
+  # @return [Number] Price to show
+  # @see https://github.com/foodcoops/foodsoft/issues/445
+  def order_article_price(order_article)
+    order_article.price.fc_price
   end
 
-  # Helper method to test pdf via rails console: OrderByGroups.new(order).save_tmp
-  def save_tmp
-    File.open("#{Rails.root}/tmp/#{self.class.to_s.underscore}.pdf", 'w') {|f| f.write(to_pdf.force_encoding("UTF-8")) }
+  def order_article_price_per_unit(order_article)
+    "#{number_to_currency(order_article_price(order_article))} / #{order_article.article.unit}"
   end
 
-  # XXX avoid underscore instead of unicode whitespace in pdf :/
-  def number_to_currency(number, options={})
-    super(number, options).gsub("\u202f", ' ') if number
-  end
+  def ordergroups(offset = nil, limit = nil)
+    result = GroupOrder.
+      ordered.
+      where(order: @orders).
+      group('groups.id').
+      offset(offset).
+      limit(limit).
+      pluck('groups.name', 'SUM(group_orders.price)', 'groups.id')
 
-  # return fontsize after scaling it with any configured factor
-  # please use this wherever you're setting a fontsize
-  def fontsize(n)
-    if FoodsoftConfig[:pdf_font_size]
-      n * FoodsoftConfig[:pdf_font_size].to_f/12
-    else
-      n
+    result.map do |item|
+      [item.first || stock_ordergroup_name] + item[1..-1]
     end
   end
 
-  # add pagebreak or vertical whitespace, depending on configuration
-  def down_or_page(space=10)
-    if @first_page
-      @first_page = false
-      return
-    end
-    if pdf_add_page_breaks?
-      start_new_page
-    else
-      move_down space
-    end
+
+  def group_order_articles(ordergroup)
+    GroupOrderArticle.
+      includes(:group_order).
+      where(group_orders: {order_id: @orders, ordergroup_id: ordergroup})
   end
 
-  protected
+  def each_order_article(&block)
+    OrderArticle.
+      ordered.
+      includes(article: :supplier).
+      includes(group_order_articles: {group_order: :ordergroup}).
+      where(order: @orders).
+      order('suppliers.name, articles.name, groups.name').
+      preload(:article_price).
+      each(&block)
+  end
```

Comment:
Can we split the scope and each? This can help plugins to add preloading for data to add to the pdf, for example.